docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        services:
+          - url: http://extension-name.com:8080
+            clusterName: kubernetes.local
+          - url: https://extension-name.ppd.cluster.k8s.local:8080
+            clusterName: admins@ppd.cluster.k8s.local
+```
+
+Note that there is an URL configuration per cluster name. The cluster
+name is extracted from the Argo CD cluster secret and must match the
+field `data.name`. In this case the UI must send a header with the
+application name so API Server can check in which cluster it should
+get the backend URL from. This will be done by inspecting the
+Application destination configuration to find the proper cluster
+name.
+
+The diagram below shows how Argo CD UI could send the request with
+the additional header to get the proxy forwarding it to the proper
+cluster:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │ HEADER: "ARGOCD-APPLICATION-NAME: ppd-application
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET https://extension-name.ppd.cluster.k8s.local:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+##### Considerations
+
+- The `idleConnTimeout` can be used to avoid accumulating too many
+  goroutines waiting slow for extensions. In this case a proper timeout
+  error (408) should be returned to the browser.
+- Headers, scheme, http verb and request body are forwarded as it is
+  received by the API server to the backend service.
+- A new header is added in the forwared request (`X-Forwarded-Host`) to
+  allow ssl redirection.
+- This proposal doesn't specify how backends should implement authz or
+  authn. This topic could be discussed as a future enhancement to the
+  proxy extension feature in Argo CD.
+
+----
+
+#### [UC-2]: As an Argo CD admin, I want to define extensions rbacs so access permissions can be enforced
+
+Extend Argo CD rbac registering a new `ResourceType` for extensions in the
+[policy configuration][3]. The current policy permission configuration is
+defined as:
+
+```
+p, <subject>, <resource>, <action>, <object>, <access>
+```
+
+With a new resource type for extensions, admins will be able to configure
+access rights per extension per project. The `object` field must contain
+the project name and the extension name in the format
+`<project>/<extension>`
+
+- **Example 1**:
+
+```
+p, role:allow-extensions, extensions, *, some-project/some-extension, allow
+```
+
+In the example 1, a permission is configured to allowing the subject
+`role:allow-extensions`, for the resource type `extensions`, for all (`*`)
+actions, in the project `some-project`, for the extension name
+`some-extension`.
+
+
+- **Example 2**:
+
+```
+p, role:allow-extensions, extensions, *, */some-extension, allow
```

Comment:
I'm a little concerned that the `<project>/<extension>` format might be confusing, since the only other slash-delimited resource names are `<project>/<application>`. Would it be acceptable to specify `<project>/<application>/<extension>`? Besides clarity, this would allow admins to further-restrict actions to particular applications if they want.

Reply:
By [reading the docs](https://argo-cd.readthedocs.io/en/stable/operator-manual/rbac/#application-resources) it felt intuitive to me to have something like:
- The `application` resource defines the action as `<project-name>/<application-name>` (already implemented)
- The `extension` resource defines the action as `<project-name>/<extension-name>`

IMHO introducing the intermediate `<application>` for the extension resource is what would make it confusing. :)

Reply:
After talking yesterday with you and @zachaller I think we identified all the RBAC format candidates:

```
p, dev, applications,                 extensions/your-name/*/path/a/b/c, default/my-application,                 allow
p, dev, extensions,                   *,                                 default/my-application/my-metrics-extn, allow
p, dev, extensions,                   my-metrics-extn/*,                 default/my-application,                 allow
p, dev, extensions/my-metrics-extn,   *,                                 default/my-application,                 allow
```

I don't think there's an obvious winner. Want to just document the options in the proposal and then decide at implementation time?

Reply:
> Want to just document the options in the proposal and then decide at implementation time?

Sounds like a good idea even if I am a bit inclined to go with the pattern from the second line. :P 

Will update the proposal with the candidates.

Reply:
My preference changes about every ten minutes, depending on how I look at it. 😆